### PR TITLE
RATEST-224:Adding Pages for report difinition workflow

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-224
This Pages will add reports using legacy UI part

### CODE REVIEW

File: ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java [hunk 1]
```diff
@@ -21,7 +21,8 @@
     private static final By MANAGE_PROVIDERS = By.cssSelector("#content a[href='/openmrs/admin/provider/index.htm']");
     private static final By MANAGE_PERSONS = By.cssSelector("#content a[href='/openmrs/admin/person/index.htm']");
     private static final By MANAGE_HTMLFORMS_PAGE = By.cssSelector("#content a[href*='/module/htmlformentry/htmlForms.list']");
-
+    private static final By REPORT_ADMINISTRATION_PAGE = By.cssSelector("#menu > li:nth-child(1) > a");
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java:24]: is `REPORT_ADMINISTRATION_PAGE `a link, button or what?

Using child(x)-value usually breaks when UI changes comes in so I would use another element reference that can hardly break.

File: ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	
+	public EditReportPage clickOnEditReportPage() {
+		clickOn(EDIT_REPORT_PAGE);
+		return new EditReportPage(this);
+	}
+	
+	@Override
+	public String getPageUrl() {
+		return "/reporting/reports/manageReports.form";
+	}
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java:23]: According to the pages i think identation is the same here

File: ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java [hunk 3]
```diff
@@ -35,6 +36,11 @@ public ManageUserPage clickOnManageUsers() {
         findElement(MANAGE_USERS).click();
         return new ManageUserPage(this);
     }
+    
+    public ManageReportsPage clickOnManageReportsPage() {
+    	clickOn(REPORT_ADMINISTRATION_PAGE);
+    	return new ManageReportsPage(this);
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java:42]: identation is the same with the same here

File: ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java:23]: Take a look at the existing page indention in [ClinicianFacingPatientDashboardPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java), [ManageAppointmentsPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/ManageAppointmentsPage.java), 
[PatientVisitsDashboardPage.](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java), [SystemAdministrationPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/SystemAdministrationPage.java) compare with the one in [EditReportPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/9b7c1c2ee3be009a5a0e8c8dcf95a2c5e713a6a3/ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java) & [ManageReportsPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/9b7c1c2ee3be009a5a0e8c8dcf95a2c5e713a6a3/ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java)

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+public class EditReportPage extends Page {
+	
+	private static final By PREVIEW_BUTTON = By.cssSelector("#previewButton");
+	private static final By START_DATE = By.cssSelector("#startDate");
+	private static final By END_DATE = By.cssSelector("#endDate");
+	private static final String startDate = "05/07/2008";
+	private static final String endDate = "20/06/2020";
+	private static final By RUN_BUTTON = By.cssSelector("#preview-parameterizable-button");
+	private static final By CLOSE_TAB = By.xpath("/html/body/div[4]/div[1]/a");
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:14]: Worth getting a better element reference for CLOSE_TAB, the current one can easily break.

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:14]: The current one is too long, using XPath will make it more clear

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:14]: @kdaud   ,Resolved them.Thanks

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:14]: > By.xpath("/html/body/div[4]/div[1]/a");
> using XPath will make it more clear

That xPath() can easily break with those divs.

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package org.openmrs.reference.page;
+
+import org.openmrs.uitestframework.page.Page;
+import org.openqa.selenium.By;
+
+public class EditReportPage extends Page {
+	
+	private static final By PREVIEW_BUTTON = By.cssSelector("#previewButton");
+	private static final By START_DATE = By.cssSelector("#startDate");
+	private static final By END_DATE = By.cssSelector("#endDate");
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:10]: Am quite not sure how more clearer that statement you would want it to look like, but i think that can be written like that even the selector can give us the clue

**Comment by kdaud** [comment]
>Take a look at the existing page indention in [ClinicianFacingPatientDashboardPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/ClinicianFacingPatientDashboardPage.java), [ManageAppointmentsPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/ManageAppointmentsPage.java), [PatientVisitsDashboardPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java), [SystemAdministrationPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/SystemAdministrationPage.java) compare with the one in [EditReportPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/9b7c1c2ee3be009a5a0e8c8dcf95a2c5e713a6a3/ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java) & [ManageReportsPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/9b7c1c2ee3be009a5a0e8c8dcf95a2c5e713a6a3/ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java)

One last comment on this PR would be revisiting this at the comment above and have this guy merged!

File: ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java:23]: Hello @kdaud all the identation in manageReportsPage is the same as module.

**Comment by sherrif10** [comment]
@kdaud  is this worth approve

File: ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package org.openmrs.reference.page;
+
+import org.openmrs.uitestframework.page.Page;
+import org.openqa.selenium.By;
+
+public class ManageReportsPage extends Page {
+	
+   private static final By EDIT_REPORT_ICON = By
+	        .cssSelector("#container > div > table > tbody > tr:nth-child(13) > td:nth-child(6) > a:nth-child(1)");
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java:9]: worth putting this on the same line

**Comment by sherrif10** [comment]
@kdaud  Addressed your comments above

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+public class EditReportPage extends Page {
+	
+   private static final By PREVIEW_BUTTON = By.cssSelector("#previewButton");
+   private static final By START_DATE_FIELD = By.cssSelector("#startDate");
+   private static final By END_DATE_FIELD = By.cssSelector("#endDate");
+   private static final By RUN_BUTTON = By.cssSelector("#preview-parameterizable-button");
+   private static final By CLOSE_TAB = By.xpath("//span[contains(text(),'close')]");
+   private static final String START_DATE = "05/07/2008";
+   private static final String END_DATE= "20/06/2020";
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:14]: worth adding a space to separate the variable name with the operator. 

Something like;
```js
private static final String END_DATE = "20/06/2020";
```

**Comment by kdaud** [comment]
One last comment would be ensuring the code indention within ManageReportsPage and EditReportPage is addressed. Take a step back [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/587#issuecomment-942368660)!

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	  clickOn(CLOSE_TAB);
+   }
+	
+   @Override
+   public String getPageUrl() {
+	  return "/reporting/reports/reportEditor.form?uuid=b39c4c4c-4881-11e7-a919-92ebcb67fe33";
+   }
+	
+}
+
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:46]: You only need one empty line at the end of the file to cater for git collaborators

**Comment by kdaud** [comment]
One last comment would be ensuring the code indention within ManageReportsPage and EditReportPage is addressed. Take a step back [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/587#issuecomment-942368660)!

**Comment by sherrif10** [comment]
Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+public class EditReportPage extends Page {
+	
+   private static final By PREVIEW_BUTTON = By.cssSelector("#previewButton");
+   private static final By START_DATE_FIELD = By.cssSelector("#startDate");
+   private static final By END_DATE_FIELD = By.cssSelector("#endDate");
+   private static final By RUN_BUTTON = By.cssSelector("#preview-parameterizable-button");
+   private static final By CLOSE_TAB = By.xpath("//span[contains(text(),'close')]");
+   private static final String START_DATE = "05/07/2008";
+   private static final String END_DATE   = "20/06/2020";
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:14]: You need only one space to separate the variable name from the operator

**Comment by kdaud** [comment]
> Resolved thanks

@sherrif10 I think you time to address the comments above;

[comment -1](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/587#discussion_r728641643)
[comment-2](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/587#discussion_r728641988)

[comment-3](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/587#issuecomment-942958438)

**Comment by sherrif10** [comment]
Resolved

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java (hunk 8 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:45]: @kdaud  you mean there is another line 46 being added ?

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:45]: > @kdaud you mean there is another line 46 being added ?

@sherrif10 does clicking at this [link](https://github.com/sherrif10/openmrs-distro-referenceapplication/edit/ratest-224/ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java?pr=%2Fopenmrs%2Fopenmrs-distro-referenceapplication%2Fpull%2F587) answers your question?

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java]: Am seeing this at a perspective of view file not edit the file, Try to click view the file, Probably that's where we are messing up from

**Comment by kdaud** [comment]
@sherrif10 when your crosscheck in your IDE, how many empty line(s) is/are with either EditReportPage or ManageReportsPage?

**Comment by sherrif10** [comment]
> how many empty line(s) is/are with either EditReportPage or ManageReportsPage?

One empty line as its showing in the committed code.

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	
+   public void clickOnPreviewButton() {
+	  clickOn(PREVIEW_BUTTON);
+   }
+	
+   public void enterStartDate(String START_DATE) {
+	  setTextToFieldNoEnter(START_DATE_FIELD, START_DATE);
+   }
+	
+   public void enterEndDate(String END_DATE) {
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:28]: even here

**Pull request closed without merging**